### PR TITLE
Fix song select realm refresh performance

### LINKED ISSUE #25824: Song Select keeps returning to currently selected map after releasing right click
### Type

Game behaviour

### Bug description

Song Select keeps returning back to the map it was currently on after I release right click which wasn't an issue before, a large stutter happens as well which wasn't as bad but still prevalent on 2023.1130.0.

### Screenshots or videos

https://github.com/ppy/osu/assets/58333920/4bec3164-6b9b-4aab-8c40-ee28afd9e697

### Version

2023.1218.0

### Logs

[shorter_runtime.log](https://github.com/ppy/osu/files/13701299/shorter_runtime.log)
[shorter_performance.log](https://github.com/ppy/osu/files/13701300/shorter_performance.log)

### PR DESCRIPTION
**Pull request by peppy**
Closes https://github.com/ppy/osu/issues/25824

Each commit is basically standalone.

If any are seen as hard-to-merge or break test expectations, we can cherry-pick the good ones for now.

87b7699fcc29ab7e70ec1f632a0341bbe08a23b9 takes things from <1 FPS to a substantially usable frame rate with occasional stutters.

The rest of the commits work to reduce the stutter down to <100ms.

Can be tested using [test database](https://drive.google.com/drive/folders/1SMSR5v2YdbKXd78JiBHrxzKnex76Tgml?usp=sharing)

### CODE REVIEW

File: osu.Game/Screens/Select/SongSelect.cs [hunk 1]
```diff
@@ -162,7 +162,7 @@ private void load(AudioManager audio, OsuColour colours, ManageCollectionsDialog
                 BleedBottom = Footer.HEIGHT,
                 SelectionChanged = updateSelectedBeatmap,
                 BeatmapSetsChanged = carouselBeatmapsLoaded,
-                FilterApplied = updateVisibleBeatmapCount,
+                FilterApplied = () => Scheduler.Add(updateVisibleBeatmapCount),
```
Comment (bdach) [inline osu.Game/Screens/Select/SongSelect.cs:165]: Why does this `Scheduler.Add()` while `carouselBeatmapsLoaded()` does `Scheduler.AddOnce()`?

File: osu.Game/Screens/Select/Carousel/CarouselGroup.cs [hunk 2]
```diff
@@ -55,6 +60,9 @@ public virtual void AddItem(CarouselItem i)
                 // criteria may be null for initial population. the filtering will be applied post-add.
                 items.Add(i);
             }
+
+            if (!i.Filtered.Value)
+                TotalItemsNotFiltered--;
```
Comment (bdach) [inline osu.Game/Screens/Select/Carousel/CarouselGroup.cs:65]:
```suggestion
            if (!i.Filtered.Value)
                TotalItemsNotFiltered++;
```

surely?

**Comment by bdach** [comment]
The invalidation logic changes are very difficult to follow. I'm still not sure whether behaviourally `BeatmapSetsLoaded` is always in the correct state now, or if `BeatmapSetsChanged` is invoked when it should be and not invoked when it shouldn't.

The test failures are not encouraging. Not sure if you want to look into them, or should I.

File: osu.Game/Screens/Select/SongSelect.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Select/SongSelect.cs:165]: Just an oversight

**Comment by peppy** [comment]
I've reverted one change for now in the interest of getting this in. Will revisit it as a separate PR.

**Comment by peppy** [comment]
One test remains flaky, trying to figure out what's going on but have not repro'd locally yet.

**Comment by bdach** [comment]
> Originally as far as I can tell it was only guarding in the realm callback, when the changeset is null. This is guaranteed by realm to only occur once, on initial event subscription.

No, `null` changeset will also happen after a realm refresh. Compare: https://github.com/ppy/osu/pull/25821.

I'm not sure what that means for this diff, since the rest of your answer is rather enigmatic, but there.

**Comment by bdach** [comment]
Tests are failing because after beatmap removal the difficulty count under search bar does not update properly anymore. It shows stale counts from before the deletion.

Previously the deletion would happen before summing up the difficulty counts, but now it happens after. So this is a regression from https://github.com/ppy/osu/pull/25838/commits/8f5d21dc703e3c1f3930e4b4e1c58c90acaba87a specifically.

**Comment by peppy** [comment]
Easy fix, luckily. Just recalculate after deletion (should have been there anyway).

If another test breaks I will not be a happy peppy.

**Comment by bdach** [comment]
Appears to be fine now, but I want this to go through a CI run with master merged before merging just to make sure I don't get clowned by unexpected crosstalk between this and the filtering fix.

**Review by bdach — APPROVED** [review]

**Pull request merged**